src/zmalloc.c

```diff
@@ -87,10 +88,36 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr, flags) je_dallocx(ptr, flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomic_fetch_add_explicit(&used_memory, (__n), memory_order_relaxed)
-#define update_zmalloc_stat_free(__n) atomic_fetch_sub_explicit(&used_memory, (__n), memory_order_relaxed)
+#if __STDC_NO_THREADS__
+#define thread_local __thread
+#else
+#include <threads.h>
+#endif
+
+/* A thread-local storage which keep the current thread's index in the used_memory_thread array. */
+static thread_local int thread_index = -1;
+/* MAX_THREADS_NUM = IO_THREADS_MAX_NUM(128) + BIO threads(3) + main thread(1). */
```

Comment:
I think this is the only hold up for me. Can we move all these settings out of `networking.c` and into a new header say `os.h`

```c
  #define IO_THREADS_MAX_NUM 128
  #ifndef CACHE_LINE_SIZE
  #if defined(__aarch64__) && defined(__APPLE__)
  #define CACHE_LINE_SIZE 128
  #else
  #define CACHE_LINE_SIZE 64
  #endif
  #endif
  ```

Reply:
Why not in config.h where we have all OS and arch specific definitions?

I'm not sure about IO_THREADS_MAX_NUM though.

Reply:
+1 on moving to config.h

>I'm not sure about IO_THREADS_MAX_NUM though.

I think this is an existing limit. As long as we single source it, it works for me.

@zuiderkwast are you suggesting we revisit this value?

Reply:
No concern with the value. I'm just not sure in which header file it should be.